terraform/modules/firewall/main.tf

```diff
@@ -0,0 +1,172 @@
+locals {
+  common_tags = {
+    terraform = "true"
+  }
+  app = var.app
+  env = var.env
+}
+
+//TODO: 
+// - Add rate limiting rule to match BCDA's config
+// - SQL Injection ruleset
+// - Cloudfront distribution (?)
+resource "aws_wafv2_web_acl" "this" {
+  name  = "${var.app}-${var.env}-wafv2-web-acl"
+  scope = "REGIONAL"
+
+  default_action {
+    allow {
+
+    }
+  }
+
+  visibility_config {
+    cloudwatch_metrics_enabled = true
+    metric_name                = "WAF_Common_Protections"
+    sampled_requests_enabled   = true
+  }
+
+  dynamic "rule" {
+    for_each = var.rate_based_rule != null ? [var.rate_based_rule] : []
+    content {
+      name     = rule.value.name
+      priority = rule.value.priority
+
+      action {
+        dynamic "count" {
+          for_each = rule.value.action == "count" ? [1] : []
+          content {}
+        }
+
+        dynamic "block" {
+          for_each = rule.value.action == "block" ? [1] : []
+          content {
+            dynamic "custom_response" {
+              for_each = rule.value.response_code != 403 ? [1] : []
+              content {
+                response_code = rule.value.response_code
+              }
+            }
+          }
+        }
+      }
+
+      statement {
+        rate_based_statement {
+          limit              = rule.value.limit
+          aggregate_key_type = "IP"
+        }
+      }
+
+      visibility_config {
+        cloudwatch_metrics_enabled = true
+        metric_name                = rule.value.name
+        sampled_requests_enabled   = true
+      }
+    }
+  }
+
+  dynamic "rule" {
+    for_each = var.ip_sets_rule
```

Comment:
I think we can make assumptions about IP sets here. If the WAF has an IP sets rule, it should be set to an empty IP set since we'll be adding ranges outside of this terraform.